TELCODOCS-405 updating multi-network policy to address SR-IOV support

kquinn1204 opened the pull request:
[TELCODOCS-405](https://issues.redhat.com//browse/TELCODOCS-405): Support Multiple Network policies for SR-IOV networks

Version(s): 4.12 and main 

Issue: https://issues.redhat.com/browse/TELCODOCS-405
Link to docs preview:

- https://51793--docspreview.netlify.app/openshift-enterprise/latest/networking/multiple_networks/configuring-multi-network-policy.html
- https://51793--docspreview.netlify.app/openshift-enterprise/latest/networking/hardware_networks/about-sriov.html
- https://51793--docspreview.netlify.app/openshift-enterprise/latest/networking/hardware_networks/about-sriov.html
Additional preview:

http://file.emea.redhat.com/kquinn/TELCODOCS-405/networking/multiple_networks/configuring-multi-network-policy.html#nw-networkpolicy-deny-all-multi-network-policy_configuring-multi-network-policy

QE review:
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->
Dev and QE LGTM added

kquinn1204 commented:
/label telco

kquinn1204 commented:
/label peer-review-needed

kquinn1204 commented:
@SchSeba @zeeke Can you please review this PR

michaelryanpeter commented:
/label peer-review-in-progress

michaelryanpeter commented on modules/nw-networkpolicy-create-cli.adoc:
```diff
@@ -116,6 +117,127 @@ spec:
++
+The use cases this example addresses are: 
+
+* Restrict traffic to a service only to other microservices that need to use it.
+* Restrict connections to a database only to the application using it.
```
With the list lead-in as a complete sentence, the current phrasing on these bullets throws me a little.

```suggestion
* Restricting the traffic to a server to only permit the microservices that need to use it.
* Restricting the connections to a database to only permit the application using it.
```
I am on the fence about my suggestion, honestly, but I hope it opens up a line of thinking.

michaelryanpeter commented on networking/hardware_networks/about-sriov.adoc:
```diff
@@ -17,6 +17,14 @@ The SR-IOV network device driver for the device determines how the VF is exposed
 
 You can use SR-IOV network devices with additional networks on your {product-title} cluster installed on bare metal or {rh-openstack-first} infrastructure for applications that require high bandwidth or low latency.
 
+You can configure multi-network policies for SR-IOV networks. The support for this is technology preview and SR-IOV additional networks are only supported with kernel NICs. They are not supported for DPDK applications.
```
Does this need a Technology Preview admonition box with all of the relevent support statements?

I assume that these are in place where they need to be (perhaps out-of-scope for this review), but I just want to double check

michaelryanpeter replied:
Unlike SR-IOV and NIC, the ISG, SSG, and repo glossary do not list _DPDK_ as an acronym that does not need to be spelled out. 
```suggestion
You can configure multi-network policies for SR-IOV networks. The support for this is technology preview and SR-IOV additional networks are only supported with kernel NICs. They are not supported for Data Plane Development Kit (DPDK) applications.
```

kquinn1204 replied:
@michaelryanpeter Yes I was not sure if tech preview snippet was required. I should probably add it as you say.

michaelryanpeter commented on networking/hardware_networks/about-sriov.adoc:
```diff
@@ -17,6 +17,14 @@ The SR-IOV network device driver for the device determines how the VF is exposed
+
+[NOTE]
+====
+Creating multi-network policies on SR-IOV networks may not deliver the same performance to applications compared to SR-IOV networks without a multi-network policy configured. 
```
```suggestion
Creating multi-network policies on SR-IOV networks might not deliver the same performance to applications compared to SR-IOV networks without a multi-network policy configured. 
```
See [ISG guidance](https://www.ibm.com/docs/en/ibm-style?topic=format-lists#o) on _may_.

michaelryanpeter commented on networking/hardware_networks/about-sriov.adoc:
```diff
@@ -17,6 +17,14 @@ The SR-IOV network device driver for the device determines how the VF is exposed
+====
+Creating multi-network policies on SR-IOV networks may not deliver the same performance to applications compared to SR-IOV networks without a multi-network policy configured. 
+====
+
```
consider removing to avoid potential whitespace/merge conflicts in the future if it does not affect your formatting.

michaelryanpeter commented on networking/multiple_networks/configuring-multi-network-policy.adoc:
```diff
@@ -10,8 +10,10 @@ As a cluster administrator, you can configure network policy for additional netw
-Other types of additional networks, such as ipvlan, are not supported.
+You can specify multi-network policy for SR-IOV and macvlan additional networks.
+Other types of additional networks, such as ipvlan, are not supported. 
+SR-IOV additional networks are technology preview and macvlan additional networks are fully supported. 
```
```suggestion
SR-IOV additional networks are a Technology Preview feature, and macvlan additional networks are fully supported. 
```
"SR-IOV additonal networks" is a little bit awkward to my ear. Is this the standard term for this feature? WDYT of "SR-IOV support for additonal networks"? It isn't as minimal, but I think flows a little better.

kquinn1204 replied:
@michaelryanpeter I believe that changes the understanding SR-IOV is the additional network type so is macvlan. These are the two additional network types supported.

michaelryanpeter commented on networking/multiple_networks/configuring-multi-network-policy.adoc:
```diff
@@ -10,8 +10,10 @@ As a cluster administrator, you can configure network policy for additional netw
 
 [NOTE]
 ====
-You can specify multi-network policy for only macvlan additional networks.
-Other types of additional networks, such as ipvlan, are not supported.
+You can specify multi-network policy for SR-IOV and macvlan additional networks.
+Other types of additional networks, such as ipvlan, are not supported. 
+SR-IOV additional networks are technology preview and macvlan additional networks are fully supported. 
+SR-IOV additional networks are only supported with kernel network interface cards (NICs). They are not supported for DPDK applications. 
```
Since there are support claims involved here, I think this makes sense as an `[IMPORTANT]` admonition. Also, I think breaking this out into an unordered list would make this easier to scan. Also, it reads more like a list than a paragraph with a thesis and supporting statements.

michaelryanpeter replied:
See the comment above about _DPDK_.

michaelryanpeter reviewed (commented):
I found some minor style and usage issues. The rest of the suggestions are potential improvements/things to consider. If you have any questions, please reach out!

michaelryanpeter commented:
/remove-label peer-review-needed
/remove-label peer-review-in-progress
/label peer-review-done

zeeke commented on modules/nw-networkpolicy-create-cli.adoc:
```diff
@@ -116,6 +117,127 @@ spec:
+ifdef::multi[]
+.Restrict traffic to a service
++
+This policy blocks traffic to pods without the `app=bookstore` label. Traffic to pods with the `app=bookstore` label is allowed. Your application could be a REST API server, marked with labels `app=bookstore` and `role=api`. 
```
I'm not sure to truly understand this phrase. The policy says that every pod with both labels `app=bookstore` and `role=api` can be accessed only by pods with label `app=bookstore`.

Pods without the `app=bookstore` label is not affected by this policy. Maybe it's what you had in mind and I didn't get it?

kquinn1204 replied:
Corrected must have messed up the logic somehow

kquinn1204 commented:
@kononovn can you review and if happy add an LGTM to this PR please.

zeeke commented on modules/nw-networkpolicy-create-cli.adoc:
```diff
@@ -116,6 +117,127 @@ spec:
+ifdef::multi[]
+.Allow traffic from external clients
++
+This policy allows external service from the public Internet directly or through a Load Balancer to access the pod. Traffic from an external service is only allowed to a pod with the label `app=web`.
```
Sorry, also this phrase seems to be not 100% coherent with the example. Pods with `app=web` can be accessed by everyone (inside and outside the cluster). In addition, an external service can connect to other pods as well. Honestly, I don't know how to implement the scenario you describe. Do you have some reference with regular NetworkPolicies doing that?

zeeke replied:
Oh, great. With the "deny-all" policy in place, the statement makes a lot of sense

kquinn1204 replied:
@zeeke I updated with a worked example end-to-end https://51793--docspreview.netlify.app/openshift-enterprise/latest/networking/multiple_networks/configuring-multi-network-policy.html#nw-networkpolicy-allow-external-clients_configuring-multi-network-policy. To be honest, I tried to configure this without much success. I think it would need a QE in case I made a error. UI would suggest just install the allow policy it's one of the canned samples. PTAL

evgenLevin commented on modules/nw-networkpolicy-allow-external-clients.adoc:
```diff
@@ -0,0 +1,100 @@
++
+[source,yaml]
+----
+apiVersion: k8s.cni.cncf.io/v1beta1
```
This policy is not going to work as expected due to the bug:
[bug](https://issues.redhat.com/browse/OCPBUGS-974)
You have to add  the PolicyType field

kquinn1204 replied:
I was pulling my hair out, wondering why it did not work. Hmm might leave out - thinking

kquinn1204 commented:
@evgenLevin Excuse my ignorance, but is VRF = Virtual routing and forwarding ? I am tidying up the section a bit to create a dedicated section on creating a default deny and then at least one other example.

zeeke commented:
> Probably we should add information that it doesn't support:
> 
> * VRF
> * IPv6
> 
> @zeeke What do you think?

@kquinn1204 yep, we are referring [Virtual Routing and Forwarding](https://www.kernel.org/doc/html/latest/networking/vrf.html). There is an open bug about MultiNetworkPolicies not working with VRF networks. I'm not sure to add a note about it on the docs, as we should provide a good description of the scenario that is not working (e.g. pods between two different VRF, inside the same VRF).

:+1: on adding a note about policies being ignored in IPv6 networks.

zeeke commented on modules/nw-networkpolicy-allow-external-clients.adoc:
```diff
@@ -0,0 +1,66 @@
+the OpenShift SDN network provider with `mode: NetworkPolicy` set.
+endif::ovn[]
+ifdef::ovn[]
+the OVN-Kubernetes network provider or the OpenShift SDN network provider with `mode: NetworkPolicy` set.
```
This phrase is already present in older version but I'm not completely sure to understand it:

> Prerequisites
> - Your cluster uses a cluster network provider that supports NetworkPolicy objects, such as the OpenShift SDN network provider with mode: NetworkPolicy set. This mode is the default for OpenShift SDN.

@dougbtv @s1061123 Can you please confirm it is still a requirement for using multinetwork-policies?

(e.g. [4.11 docs multinetwork-policy](https://docs.openshift.com/container-platform/4.11/networking/multiple_networks/configuring-multi-network-policy.html))

Maybe we can fix it before duplicating, if needed

kquinn1204 replied:
@zeeke Yes I am building on previous version so I am not 100% sure if this statement is still relevant. The source code makes a distinction between ovn and sdn producing different output depending on the context.

kquinn1204 commented:
@zeeke @dougbtv @s1061123  I wonder working with this network policy feature has there ever been a request to  **disable** the created policy. In that way it still exists, and you could have lots configured and simply turn on when needed. Deleting it removes it, and you have to recreate from scratch.

kquinn1204 commented:
/label peer-review-needed been through one round of peer review, but I added additional examples

kquinn1204 commented:
/label peer-review-needed

sheriff-rh commented on networking/multiple_networks/configuring-multi-network-policy.adoc:
```diff
@@ -6,14 +6,22 @@ include::_attributes/common-attributes.adoc[]
 toc::[]
 
-As a cluster administrator, you can configure network policy for additional networks.
+As a cluster administrator, you can configure multi-network for additional networks. You can specify multi-network policy for SR-IOV and macvlan additional networks. Macvlan additional networks are fully supported other types of additional networks, such as ipvlan, are not supported.
```
I think you need a period here: 

"Macvlan addditional networks are fully supported. Other types of ..."

sheriff-rh commented on networking/multiple_networks/configuring-multi-network-policy.adoc:
```diff
@@ -6,14 +6,22 @@ include::_attributes/common-attributes.adoc[]
+
+[IMPORTANT]
+====
+Support for configuring multi-network policies for SR-IOV additional networks is a Technology Preview feature and:
```
I think you are missing the word "is" at the end. Otherwise it reads "TP feature and only supported..."

I would just do sentences here for clarity:

"Support for configuring multi-network policies for SR-IOV additional networks is a Technology Preview feature and is only supported with kernel network interface cards (NICs). SR-IOV is not supported for Data Plane Development Kit (DPDK) applications."

This would be a great opportunity to `link:` to the [technology preview support scope](https://access.redhat.com/support/offerings/techpreview/).

sheriff-rh reviewed (changes requested):
This looks really good! I have a few nits and a few broader suggestions as well. Please reach out if you have any questions! 

I believe your ifdef/ifeval logic is correct, but some of it cannot be confirmed unless checked against several branches. Just be sure to follow up after merging and verify everything looks as expected!

sheriff-rh commented:
/remove-label peer-review-in-progress
/label peer-review-done

sheriff-rh commented on networking/hardware_networks/about-sriov.adoc:
```diff
@@ -17,6 +17,23 @@ The SR-IOV network device driver for the device determines how the VF is exposed
+For more information about the support scope of Red Hat Technology Preview
+features, see https://access.redhat.com/support/offerings/techpreview/.
+====
+
```
I suspected it was a complex situation. Thanks!

kalexand-rh replied:
You can (and should) still use the snippet.

kquinn1204 commented:
@zeeke @evgenLevin I would love to get this merged before the weekend if possible. Do you think it is now OK for a LGTM ? I added examples and 1 small diagram

zeeke commented:
LGTM

Very nice and effective  diagram, and really appreciate the "allow from namespace" chapter :+1:

evgenLevin commented:
LGTM

kquinn1204 commented:
/label merge-review-needed

stevsmit commented on modules/nw-networkpolicy-create-cli.adoc:
```diff
@@ -121,6 +124,95 @@ where
+--
+where
+
+`<network_name>`:: Specifies the name of a network attachment definition.
```
I'm not sure, but I don't think this is rendering correctly. Also specified is capitalized and I don't think it should be.

kalexand-rh replied:
"Specifies" should be capitalized.

stevsmit commented:
@kquinn1204 A few minor things to consider before merging. Most aren't deal breaking. The biggest concern is the rendering issue I pointed out. Thanks.

kquinn1204 commented on modules/nw-networkpolicy-allow-application-all-namespaces.adoc:
```diff
@@ -0,0 +1,148 @@
+$ oc run test-$RANDOM --namespace=secondary --rm -i -t --image=alpine -- sh
+----
+
+. In your shell, run the following command and observe the request is allowed:
```
changed to "observe that the request is allowed"

kquinn1204 commented on modules/nw-networkpolicy-create-cli.adoc:
```diff
@@ -121,6 +124,95 @@ where
+
+where
+
+`<network_name>`:: Specifies the name of a network attachment definition.
```
I inherited this from a previous revision and simply copied to a new section. I think this was how it is supposed to render and to be honest I had a difficult time getting this `ifdef` logic to work as expected, so I am not really keen on messing around with it more. Hopefully we can live with it

kquinn1204 commented:
@stevsmit I have addressed comments above bar one which I elaborated on above. I spent a few cycles trying to get the idef stuff to work correctly, and I decided to stick with the structure I had inherited in these sections

kquinn1204 commented:
/label merge-review-needed

kquinn1204 commented:
/remove-label merge-review-needed

kalexand-rh commented on modules/nw-networkpolicy-create-cli.adoc:
```diff
@@ -121,6 +124,95 @@ where
+ifdef::multi[]
++
+--
+where
```
```suggestion
where:
```
Per the [contrib guide](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output).

kquinn1204 replied:
Updated

kalexand-rh commented on modules/nw-networkpolicy-deny-all-allowed.adoc:
```diff
@@ -0,0 +1,93 @@
+--
+where
+
+`<network_name>` specifies the name of a network attachment definition.
```
This is the one that's not rendering right. You either need all of the annotations to be in the definition list or all of them to use callouts.

kalexand-rh replied:
You could move them all into the definition list.

kquinn1204 replied:
I wrapped in more expansive ifdef statements

kalexand-rh commented on modules/nw-networkpolicy-allow-application-particular-namespace.adoc:
```diff
@@ -0,0 +1,198 @@
+If you log in with a user with the `cluster-admin` role, then you can create a network policy in any namespace in the cluster.
+====
+
+Follow this procedure to configure a policy that allows traffic to a pod with the label `app=web` from a particular namespace. You might want to do this to:
```
Something like: 
```suggestion
You can configure a policy that allows traffic to a pod with the label `app=web` from a particular namespace to complete the following tasks:
```

kalexand-rh reviewed (commented):
I agree with Steven - one of the annotation sets needs to be changed.

kalexand-rh commented:
@kquinn1204, you need to squash before we can merge.

bergerhoffer reviewed (commented):
Checked just the things since the last merge review, and the most important things seem to have been addressed. Merging!

bergerhoffer commented:
/cherrypick enterprise-4.12
